### PR TITLE
Always allocate an exception object at throw site

### PR DESCRIPTION
**Pull request by smowton**
Previously we could erroneously fail to initialise the object on second
and subsequent exception throw-sites in a particular function, and so throw
a nondet object rather than an initialised exception instance. The attached
test case checks for this by ensuring that in particular we can't throw null
and so hit an assertion when trying to read its class-identifier for catch dispatch.

### CODE REVIEW

**Comment by smowton** [comment]
Companion: https://github.com/diffblue/test-gen/pull/880

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 1]
```diff
@@ -90,47 +90,46 @@ codet java_bytecode_instrumentt::throw_exception(
...
+  code_assignt assign_new(new_symbol.symbol_expr(), new_instance);
 
   side_effect_expr_throwt throw_expr;
-  throw_expr.add_source_location()=original_loc;
-  throw_expr.move_to_operands(exc);
+  throw_expr.copy_to_operands(new_symbol.symbol_expr());
+
+  code_blockt throw_code;
+  throw_code.move_to_operands(assign_new);
+  throw_code.copy_to_operands(code_expressiont(throw_expr));
```
Comment (allredj) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Just for my own culture: Why can't we not use `move_to_operands` here?

Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Move-to-operands takes a mutable ref (an `exprt&`) not an `exprt&&`, which cannot be taken to an rvalue like this.

Comment (allredj) [inline src/java_bytecode/java_bytecode_instrument.cpp]: OK thanks!

**Pull request merged**